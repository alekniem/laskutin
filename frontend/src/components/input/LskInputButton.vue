<template>
    <v-btn :color="color" :disabled="disabled" :id="cid" depressed small>
        {{ buttonText | trans }}
    </v-btn>
</template>

<script>
export default {
    name: "LskInputButton",

    props: {
        buttonText: { type: String, required: true },
        cid: { type: String, required: true },
        color: { type: String, required: false, default: "primary" },
        disabled: { type: Boolean, required: false, default: false },
    },
};
</script>
