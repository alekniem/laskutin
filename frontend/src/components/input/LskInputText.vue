<template>
    <ValidationProvider
        :customMessages="{ required: trans('Required') }"
        :vid="vid"
        ref="provider"
        rules="required"
        tag="div"
        #default="{ errors, valid }"
    >
        <v-text-field
            @change="$emit('input', $event)"
            :error-messages="errors"
            :id="cid"
            :label="trans(label)"
            :success="valid"
            :value="value"
        />
    </ValidationProvider>
</template>

<script>
import { ValidationProvider } from "vee-validate";

export default {
    name: "LskInputText",

    components: {
        ValidationProvider,
    },

    props: {
        cid: { type: String, required: true },
        label: { type: String, required: true },
        value: { type: [String, Number], required: true },
        vid: { type: String, required: true },
    },
};
</script>
