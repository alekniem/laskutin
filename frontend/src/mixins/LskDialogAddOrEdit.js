import lsk_api from "@/plugins/lsk_api";
import { ValidationObserver } from "vee-validate";

export default {
    components: {
        ValidationObserver,
    },

    model: {
        prop: "showDialog",
        event: "dialog",
    },

    props: {
        id: { type: [String, Number], required: false, default: null },
        currentData: { type: Object, required: false, default: null },
        showDialog: { type: Boolean, required: true },
    },

    data() {
        return {
            dialogVisible: this.showDialog,
            fields: [],
            inputData: {},
            errors: {},
            otherErrors: [],
        };
    },

    computed: {
        apiActionName() {
            return null;
        },

        originalData() {
            if (this.currentData === null) {
                return this.emptyData();
            }
            return this.currentData;
        },

        saveButtonDisabled() {
            for (let key in this.inputData) {
                if (this.hasField(this.originalData, key) && this.inputData[key] != this.originalData[key]) {
                    return false;
                }
            }
            return true;
        },
    },

    watch: {
        showDialog(value) {
            this.dialogVisible = value;
        },
    },

    created() {
        this.resetInputData();
        this.resetErrors();
    },

    methods: {
        emptyData() {
            let result = {};
            this.fields.forEach((field) => {
                result[field] = "";
            });
            return result;
        },

        resetInputData() {
            this.inputData = {};
            this.fields.forEach((field) => {
                if (this.hasField(this.originalData, field)) {
                    this.$set(this.inputData, field, this.originalData[field]);
                } else {
                    this.$set(this.inputData, field, "");
                }
            });
        },

        resetErrors() {
            this.errors = {};
            this.fields.forEach((field) => this.$set(this.errors, field, []));
            this.otherErrors = [];
        },

        processCancel() {
            this.$emit("dialog", false);
        },

        processSave() {
            lsk_api
                .query(this.apiActionName, this.id, this.inputData)
                .then((response) => this.parseResponse(response.data))
                .catch((error) => this.parseError(error.response.data));
        },

        parseResponse() {
            this.$emit("dialog", false);
        },

        parseError(data) {
            this.resetErrors();

            if (this.hasField(data, "errors")) {
                for (let id in data.errors) {
                    for (let key in data.errors[id]) {
                        if (this.hasField(this.errors, id)) {
                            this.errors[id].push(this.trans(data.errors[id][key]));
                        } else {
                            this.otherErrors.push(this.trans(data.errors[id][key]));
                        }
                    }
                }
            } else if (this.hasField(data, "message")) {
                this.otherErrors.push(this.trans(data.message));
            } else {
                this.otherErrors.push(this.trans("Unidentified error"));
            }

            this.$refs.observer.setErrors(this.errors);
        },
    },
};
