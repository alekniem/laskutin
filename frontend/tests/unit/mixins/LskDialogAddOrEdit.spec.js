import axios from "axios";
import flushPromises from "flush-promises";
import LskInputButton from "@/components/input/LskInputButton";
import LskInputText from "@/components/input/LskInputText";
import LskDialogAddOrEdit from "@/mixins/LskDialogAddOrEdit";
import { createWrapper } from "UT/lsk_test_helper";

jest.mock("axios");

axios.post.mockImplementation((url, data) => {
    if (url === "customer" && data.test_one === "TEST ONE") {
        return Promise.resolve({
            data: {
                message: "POST QUERY OK",
            },
        });
    }
    return Promise.reject({
        response: {
            data: {
                message: "POST QUERY ERROR",
            },
        },
    });
});

const testComponent = function() {
    return {
        name: "TestComponent",
        components: {
            LskInputButton,
            LskInputText,
        },
        mixins: [LskDialogAddOrEdit],
        data() {
            return {
                fields: ["test_one", "test_two"],
            };
        },
        computed: {
            apiActionName() {
                return "customerStore";
            },
        },
        template: `
            <v-dialog max-width="700" persistent v-model="dialogVisible">
                <ValidationObserver #default="{ handleSubmit, invalid }" ref="observer">
                    <v-form>
                        <LskInputText
                            cid="testOneInput"
                            label="Test one"
                            vid="test_one"
                            v-model="inputData.test_one"
                        />
                        <LskInputText
                            cid="testTwoInput"
                            label="Test two"
                            vid="test_two"
                            v-model="inputData.test_two"
                        />
                        <LskInputButton
                            @click.native="processCancel"
                            buttonText="Cancel"
                            cid="cancelButton"
                        />
                        <LskInputButton
                            @click.native="handleSubmit(processSave)"
                            :disabled="invalid || saveButtonDisabled"
                            buttonText="Save"
                            cid="saveButton"
                        />
                    </v-form>
                </ValidationObserver>
            </v-dialog>
        `,
    };
};

describe("LskDialogAddOrEdit", () => {
    it("should match snapshot", async () => {
        const wrapper = createWrapper(testComponent(), {
            propsData: {
                showDialog: false,
            },
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog not visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(testComponent(), {
            propsData: {
                showDialog: true,
            },
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(testComponent(), {
            propsData: {
                showDialog: true,
            },
        });
        await flushPromises();
        wrapper.vm._data.inputData.test_one = "";
        wrapper.vm._data.inputData.test_two = "";
        await flushPromises();
        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("error messages");
    });

    it("should have correct input data", async () => {
        const wrapper = createWrapper(testComponent(), {
            propsData: {
                currentData: {
                    test_one: "TEST ONE",
                    test_two: "TEST TWO",
                },
                showDialog: true,
            },
        });
        await flushPromises();
        expect(wrapper.vm._data.inputData).toStrictEqual(wrapper.vm._props.currentData);
    });

    it("should change 'save' button from disabled to enabled", async () => {
        const wrapper = createWrapper(testComponent(), {
            propsData: {
                showDialog: true,
            },
        });
        await flushPromises();

        wrapper.vm._data.inputData.test_one = "";
        wrapper.vm._data.inputData.test_two = "";
        await flushPromises();
        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.find("#saveButton").attributes().disabled).toBeDefined();

        wrapper.vm._data.inputData.test_one = "TEST ONE";
        wrapper.vm._data.inputData.test_two = "TEST TWO";
        await flushPromises();
        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.find("#saveButton").attributes().disabled).toBeUndefined();
    });

    it("should emit 'dialog' event on success", async () => {
        const wrapper = createWrapper(testComponent(), {
            propsData: {
                showDialog: true,
            },
        });
        await flushPromises();
        wrapper.vm._data.inputData.test_one = "TEST ONE";
        wrapper.vm._data.inputData.test_two = "TEST TWO";
        await flushPromises();
        wrapper.find("#saveButton").trigger("click");
        await flushPromises();
        let key = axios.post.mock.calls.length - 1;
        expect(axios.post.mock.calls[key][1]).toStrictEqual(wrapper.vm._data.inputData);
        expect(wrapper.emitted().dialog).toStrictEqual([[false]]);
    });

    it("should show an error message on failure", async () => {
        const wrapper = createWrapper(testComponent(), {
            propsData: {
                showDialog: true,
            },
        });
        await flushPromises();
        wrapper.vm._data.inputData.test_one = "INVALID";
        wrapper.vm._data.inputData.test_two = "INVALID";
        await flushPromises();
        wrapper.find("#saveButton").trigger("click");
        await flushPromises();
        let key = axios.post.mock.calls.length - 1;
        expect(axios.post.mock.calls[key][1]).toStrictEqual(wrapper.vm._data.inputData);
        expect(wrapper.vm._data.otherErrors).toStrictEqual(["POST QUERY ERROR"]);
    });

    it("should emit 'dialog' event on cancel", async () => {
        const wrapper = createWrapper(testComponent(), {
            propsData: {
                showDialog: true,
            },
        });
        await flushPromises();
        wrapper.find("#cancelButton").trigger("click");
        await flushPromises();
        expect(wrapper.emitted().dialog).toStrictEqual([[false]]);
    });
});
