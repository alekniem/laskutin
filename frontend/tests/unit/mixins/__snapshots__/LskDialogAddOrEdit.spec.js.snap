// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LskDialogAddOrEdit should match snapshot: dialog not visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
</div>
`;

exports[`LskDialogAddOrEdit should match snapshot: dialog visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div>
              <div class="v-input theme--light v-text-field v-text-field--is-booted">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot">
                      <label for="testOneInput"
                             class="v-label theme--light"
                             style="left: 0px; position: absolute;"
                      >
                        Test one
                      </label>
                      <input id="testOneInput"
                             type="text"
                      >
                    </div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div"
                                             name="message-transition"
                                             class="v-messages__wrapper"
                      >
                      </transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div class="v-input theme--light v-text-field v-text-field--is-booted">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot">
                      <label for="testTwoInput"
                             class="v-label theme--light"
                             style="left: 0px; position: absolute;"
                      >
                        Test two
                      </label>
                      <input id="testTwoInput"
                             type="text"
                      >
                    </div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div"
                                             name="message-transition"
                                             class="v-messages__wrapper"
                      >
                      </transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <button type="button"
                    class="v-btn v-btn--has-bg theme--light v-size--small primary"
                    id="cancelButton"
            >
              <span class="v-btn__content">
                Cancel
              </span>
            </button>
            <button type="button"
                    disabled="disabled"
                    class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                    id="saveButton"
            >
              <span class="v-btn__content">
                Save
              </span>
            </button>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`LskDialogAddOrEdit should match snapshot: error messages 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div>
              <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted error--text">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot">
                      <label for="testOneInput"
                             class="v-label theme--light error--text"
                             style="left: 0px; position: absolute;"
                      >
                        Test one
                      </label>
                      <input id="testOneInput"
                             type="text"
                      >
                    </div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light error--text"
                         role="alert"
                    >
                      <transition-group-stub tag="div"
                                             name="message-transition"
                                             class="v-messages__wrapper"
                      >
                        <div class="v-messages__message">
                          Required
                        </div>
                      </transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted error--text">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot">
                      <label for="testTwoInput"
                             class="v-label theme--light error--text"
                             style="left: 0px; position: absolute;"
                      >
                        Test two
                      </label>
                      <input id="testTwoInput"
                             type="text"
                      >
                    </div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light error--text"
                         role="alert"
                    >
                      <transition-group-stub tag="div"
                                             name="message-transition"
                                             class="v-messages__wrapper"
                      >
                        <div class="v-messages__message">
                          Required
                        </div>
                      </transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <button type="button"
                    class="v-btn v-btn--has-bg theme--light v-size--small primary"
                    id="cancelButton"
            >
              <span class="v-btn__content">
                Cancel
              </span>
            </button>
            <button type="button"
                    disabled="disabled"
                    class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                    id="saveButton"
            >
              <span class="v-btn__content">
                Save
              </span>
            </button>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;
